src/limited.rs

```diff
@@ -0,0 +1,308 @@
+use crate::{Body, SizeHint};
+use bytes::Buf;
+use http::HeaderMap;
+use std::error::Error;
+use std::fmt;
+use std::pin::Pin;
+use std::task::{Context, Poll};
+
+/// A length limited body.
+///
+/// This body will return an error if more than the configured number
+/// of bytes are returned on polling the wrapped body.
+#[derive(Clone, Copy, Debug)]
+pub struct Limited<B> {
+    remaining: usize,
+    inner: B,
+}
+
+impl<B> Limited<B> {
+    /// Create a new `Limited`.
+    pub fn new(inner: B, limit: usize) -> Self {
+        Self {
+            remaining: limit,
+            inner,
+        }
+    }
+}
+
+impl<B> Body for Limited<B>
+where
+    B: Body + Unpin,
+{
+    type Data = B::Data;
+    type Error = LengthLimitError<B::Error>;
```

Comment:
I think this should be `BoxError` for the same reason that tower uses `BoxError`. Otherwise if you're composing multiple bodies you get a huge enum thats very hard to `match` on, and changing the order of layers breaks that `match` since its nests the enums differently.